neqo-http3/src/connection.rs

```diff
@@ -520,37 +517,99 @@ impl Http3Connection {
             return Err(Error::InvalidStreamId);
         }
-        self.send_streams.remove(&stream_id);
-        if let Some(mut s) = self.recv_streams.remove(&stream_id) {
-            s.reset(error, ResetType::App)?;
+
+        if let Some(mut s) = self.send_streams.remove(&stream_id) {
+            s.stop_sending(CloseType::ResetApp(error));
         }
+        conn.stream_reset_send(stream_id, error)?;
+        Ok(())
+    }
+
+    pub fn stream_stop_sending(
+        &mut self,
+        conn: &mut Connection,
+        stream_id: u64,
+        error: AppError,
+    ) -> Res<()> {
+        qinfo!(
+            [self],
+            "Send stop sending for stream {} error={}.",
+            stream_id,
+            error
+        );
+        if self.recv_stream_is_critical(stream_id) {
+            return Err(Error::InvalidStreamId);
+        }
+
+        self.recv_streams
+            .remove(&stream_id)
+            .map_or(Ok(()), |mut s| s.reset(CloseType::ResetApp(error)))?;
 
         // Stream may be already be closed and we may get an error here, but we do not care.
-        mem::drop(conn.stream_reset_send(stream_id, error));
-        // Stream may be already be closed and we may get an error here, but we do not care.
-        mem::drop(conn.stream_stop_sending(stream_id, error));
+        conn.stream_stop_sending(stream_id, error)?;
+        Ok(())
+    }
+
+    pub fn cancel_http_request(
+        &mut self,
+        stream_id: u64,
+        error: AppError,
+        conn: &mut Connection,
+    ) -> Res<()> {
+        qinfo!([self], "reset_:stream {} error={}.", stream_id, error);
+        let send_stream = self.send_streams.get(&stream_id);
+        let recv_stream = self.recv_streams.get(&stream_id);
+        match (send_stream, recv_stream) {
+            (None, None) => return Err(Error::InvalidStreamId),
+            (Some(s), None) => {
+                if !matches!(s.stream_type(), Http3StreamType::Http) {
+                    return Err(Error::InvalidStreamId);
+                }
+                // Stream may be already be closed and we may get an error here, but we do not care.
+                mem::drop(self.stream_reset_send(conn, stream_id, error));
+            }
+            (None, Some(s)) => {
+                if !matches!(
+                    s.stream_type(),
+                    Http3StreamType::Http | Http3StreamType::Push
+                ) {
+                    return Err(Error::InvalidStreamId);
+                }
+
+                // Stream may be already be closed and we may get an error here, but we do not care.
+                mem::drop(self.stream_stop_sending(conn, stream_id, error));
+            }
+            (Some(s), Some(r)) => {
+                let st = s.stream_type();
+                let rt = r.stream_type();
+                if st != rt {
+                    panic!("The sender and receive stream types must be the same");
+                }
```

Comment:
`debug_assert_eq!(s.stream_type(), r.stream_type());` should be enough here.  This is not a problem we would need to check at runtime in the general case.  If this has ever happened to you, then a runtime (Internal) error might be OK, but I am guessing that we can't get into this state.

Reply:
No we cannot get into this state at all.